From 703b0d3efe1c2ba9e5328bbece6d0c6027a45f3b Mon Sep 17 00:00:00 2001
From: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
Date: Mon, 22 Apr 2024 13:23:25 +0300
Subject: [PATCH 016/104] Bluetooth: hci_sync: Fix BLE devices turning wakeup
 on/off

This is part 1 from initial patch provided by
  "zhongjun.yu" <zhongjun.yu@quectel.com>
and ported to v6.5:

1) Fixed an issue where BLE devices could not turn wakeup on or off.
2) Resolved an issue where bluez lib could not receive
   MGMT_EV_DEVICE_FLAGS_CHANGED notifications after sending
   MGMT_OP_SET_DEVICE_FLAGS

<Change Type>: Develop the function to set the Bluetooth device wake-up
switch

<<< Test Notes >>>
<Test-Proposal>:
1) python script set_wakeup.py
2) valve sw guys Developing UI validation
<Stress-Test>: Y

[fwd-ported to v6.11 and dropped part 2 due to a breakage with BlueZ 5.78+]
Link: https://gitlab.steamos.cloud/holo-team/tasks/-/issues/1456
Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
---
 net/bluetooth/hci_sync.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/net/bluetooth/hci_sync.c b/net/bluetooth/hci_sync.c
index 589aba62a..3d724479d 100644
--- a/net/bluetooth/hci_sync.c
+++ b/net/bluetooth/hci_sync.c
@@ -2749,6 +2749,7 @@ static u8 hci_update_accept_list_sync(struct hci_dev *hdev)
 	struct bdaddr_list *b, *t;
 	u8 num_entries = 0;
 	bool pend_conn, pend_report;
+	struct hci_conn_params *conn_params;
 	u8 filter_policy;
 	size_t i, n;
 	int err;
@@ -2826,6 +2827,15 @@ static u8 hci_update_accept_list_sync(struct hci_dev *hdev)
 			continue;
 		}
 
+		/* During suspend, only wakeable devices can be in acceptlist */
+		conn_params = hci_conn_params_lookup(hdev,&b->bdaddr,b->bdaddr_type);
+		if (conn_params && hdev->suspended &&
+		    !(conn_params->flags & HCI_CONN_FLAG_REMOTE_WAKEUP)) {
+			hci_le_del_accept_list_sync(hdev, &b->bdaddr,
+						    b->bdaddr_type);
+			continue;
+		}
+
 		num_entries++;
 	}
 
-- 
2.52.0

